File: dotnet/CalBot.CSharp/ArithmeticVerifier.cs
namespace CalBot.CSharp;

/// <summary>
/// C# Arithmetic Verifier for CalBot.
///
/// This class exists because the calculator app needed another
/// verification path, and the .NET runtime was already being
/// shipped for F# and VB.NET, so we might as well use all three
/// CLR languages to verify that 2+2=4.
///
/// Enterprise design patterns applied:
/// - Strategy pattern (the operation)
/// - Factory method (CreateVerifier)
/// - Result object (VerificationResult)
/// - Async/await (because it's C#)
/// - LINQ (because it's C#)
/// - Nullable reference types (because it's 2026)
/// </summary>
public static class ArithmeticVerifier
{
    public record VerificationResult(
        int Operand1,
        int Operand2,
        string Operator,
        int ExpectedResult,
        int ActualResult,
        bool Verified,
        string Language = "C# 13 (.NET 10)"
    );

    /// <summary>
    /// Verify an arithmetic operation using C#'s checked arithmetic.
    /// The 'checked' context ensures overflow throws instead of wrapping,
    /// because we take integer safety seriously in this calculator app.
    /// </summary>
    public static VerificationResult Verify(int a, int b, string op)
    {
        int result = checked(op switch
        {
            "+" => a + b,
            "-" => a - b,
            "*" => a * b,
            "/" when b != 0 => a / b,
            "/" => throw new DivideByZeroException(
                "C# refuses to divide by zero for your calculator."),
            _ => throw new ArgumentException(
                $"Unsupported operator: {op}")
        });

        return new VerificationResult(
            Operand1: a,
            Operand2: b,
            Operator: op,
            ExpectedResult: result,
            ActualResult: result,
            Verified: true
        );
    }

    /// <summary>
    /// Cross-check: given an expected result, verify it matches C#'s computation.
    /// This is the entry point called from the Android JNI bridge.
    /// </summary>
    public static VerificationResult CrossCheck(int a, int b, string op, int expected)
    {
        var result = Verify(a, b, op);
        return result with
        {
            ExpectedResult = expected,
            Verified = result.ActualResult == expected
        };
    }

    /// <summary>
    /// Bulk verification using LINQ because this is C# and we have standards.
    /// </summary>
    public static IEnumerable<VerificationResult> VerifyBatch(
        IEnumerable<(int a, int b, string op, int expected)> cases)
        => cases.Select(c => CrossCheck(c.a, c.b, c.op, c.expected));
}

File: dotnet/CalBot.VBNet/ArithmeticVerifier.vb
Namespace Global.CalBot.VBNet

    ''' <summary>
    ''' VB.NET Arithmetic Verifier for CalBot.
    '''
    ''' Visual Basic .NET: the language that enterprise developers
    ''' used to build CRUD apps in 2003 is now verifying blockchain
    ''' arithmetic on an Android phone in 2026.
    '''
    ''' VB.NET was chosen because:
    ''' 1. It adds to the language count
    ''' 2. The .NET runtime was already being shipped for F# and C#
    ''' 3. It would be rude to leave VB out of a .NET deployment
    ''' 4. BASIC (line 20 GOTO 10) needed a modern descendant in the project
    '''
    ''' This is the most verbose arithmetic verifier in the codebase,
    ''' which is appropriate for a language that uses "End If" instead of "}".
    ''' </summary>
    Public Class ArithmeticVerifier

        Public Class VerificationResult
            Public Property Operand1 As Integer
            Public Property Operand2 As Integer
            Public Property [Operator] As String
            Public Property ExpectedResult As Integer
            Public Property ActualResult As Integer
            Public Property Verified As Boolean
            Public Property Language As String = "VB.NET 16 (.NET 10)"

            Public Overrides Function ToString() As String
                Dim verdict As String = If(Verified, "PASS", "FAIL")
                Return $"VB.NET: {Operand1} {[Operator]} {Operand2} = {ActualResult} (expected {ExpectedResult}) [{verdict}]"
            End Function
        End Class

        ''' <summary>
        ''' Verify an arithmetic operation using VB.NET's arithmetic operators.
        ''' Uses Select Case because VB.NET has standards.
        ''' </summary>
        Public Shared Function Verify(a As Integer, b As Integer, op As String) As VerificationResult
            Dim result As Integer

            Select Case op
                Case "+"
                    result = a + b
                Case "-"
                    result = a - b
                Case "*"
                    result = a * b
                Case "/"
                    If b = 0 Then
                        Throw New DivideByZeroException(
                            "VB.NET refuses to divide by zero. " &
                            "This is not a drill. " &
                            "Contact your nearest COBOL programmer.")
                    End If
                    result = a \ b  ' Integer division in VB.NET uses \
                Case Else
                    Throw New ArgumentException(
                        $"Unsupported operator: {op}. " &
                        "VB.NET only does +, -, *, /. " &
                        "For anything else, use F#.")
            End Select

            Dim verification As New VerificationResult()
            verification.Operand1 = a
            verification.Operand2 = b
            verification.Operator = op
            verification.ExpectedResult = result
            verification.ActualResult = result
            verification.Verified = True
            Return verification
        End Function

        ''' <summary>
        ''' Cross-check an expected result against VB.NET's computation.
        ''' This is the entry point called from the Android JNI bridge,
        ''' which calls C, which calls .NET, which calls VB.NET,
        ''' to verify arithmetic that Rust already computed correctly.
        ''' </summary>
        Public Shared Function CrossCheck(a As Integer, b As Integer,
                                           op As String,
                                           expected As Integer) As VerificationResult
            Dim result = Verify(a, b, op)
            result.ExpectedResult = expected
            result.Verified = (result.ActualResult = expected)
            Return result
        End Function

        ''' <summary>
        ''' Fibonacci in VB.NET because every language in this project
        ''' implements Fibonacci. It's basically a rite of passage.
        ''' </summary>
        Public Shared Function Fibonacci(n As Integer) As Long
            If n <= 0 Then Return 0
            If n = 1 Then Return 1

            Dim a As Long = 0
            Dim b As Long = 1
            For i As Integer = 2 To n
                Dim temp As Long = a + b
                a = b
                b = temp
            Next
            Return b
        End Function

    End Class

End Namespace
